### ISSUE TITLE
redis example fails to build

### ISSUE THREAD

**Issue by englishm** [hashicorp/packer#41]
with error: "Unable to locate package redis-server"

`example.json` has these contents:

```
{
  "builders": [{
    "type": "amazon-ebs",
    "access_key": "REDACTED",
    "secret_key": "REDACTED",
    "region": "us-east-1",
    "source_ami": "ami-de0d9eb7",
    "instance_type": "t1.micro",
    "ssh_username": "ubuntu",
    "ami_name": "packer-example {{.CreateTime}}"
  }],
  "provisioners": [{
    "type": "shell",
    "inline": [
      "sudo apt-get update",
      "sudo apt-get install -y redis-server"
    ]
  }]
}

```

Running the commands from the [getting started guide](http://www.packer.io/intro/getting-started/provision.html):

```
$ packer validate example.json
Template validated successfully.
$ packer build example.json
amazon-ebs output will be in this color.

==> amazon-ebs: Creating temporary keypair for this instance...
==> amazon-ebs: Creating temporary security group for this instance...
==> amazon-ebs: Authorizing SSH access on the temporary security group...
==> amazon-ebs: Launching a source AWS instance...
==> amazon-ebs: Waiting for instance to become ready...
==> amazon-ebs: Connecting to the instance via SSH...
==> amazon-ebs: Provisioning with shell script: /var/folders/gp/gn1043nx4fn5k3db55x6w9l40000gn/T/packer-shell151021730
Ign http://security.ubuntu.com precise-security InRelease
Ign http://archive.ubuntu.com precise InRelease
Get:1 http://security.ubuntu.com precise-security Release.gpg [198 B]
Ign http://archive.ubuntu.com precise-updates InRelease
Get:2 http://security.ubuntu.com precise-security Release [49.6 kB]
Hit http://archive.ubuntu.com precise Release.gpgse 0 B
Get:3 http://archive.ubuntu.com precise-updates Release.gpg [198 B]
Hit http://archive.ubuntu.com precise Release0%] [2 Rel
Get:4 http://archive.ubuntu.com precise-updates Release [49.6 kB]
Get:5 http://security.ubuntu.com precise-security/main amd64 Packages [295 kB]
Hit http://archive.ubuntu.com precise/main amd64 Packages
Hit http://archive.ubuntu.com precise/restricted amd64 Packages
Hit http://archive.ubuntu.com precise/universe amd64 Packages
Hit http://archive.ubuntu.com precise/multiverse amd64 Packages
Hit http://archive.ubuntu.com precise/main i386 Packages
Hit http://archive.ubuntu.com precise/restricted i386 Packages
Hit http://archive.ubuntu.com precise/universe i386 Packages
    amazon-ebs: Get:6 http://security.ubuntu.com precise-security/restricted amd64 Packages [4627 B]
Hit http://archive.ubuntu.com precise/multiverse i386 Packages
Get:7 http://security.ubuntu.com precise-security/universe amd64 Packages [77.1 kB]
Hit http://archive.ubuntu.com precise/main TranslationIndex
Hit http://archive.ubuntu.com precise/multiverse TranslationIndex
Get:8 http://security.ubuntu.com precise-security/multiverse amd64 Packages [2182 B]
Get:9 http://security.ubuntu.com precise-security/main i386 Packages [309 kB]
Hit http://archive.ubuntu.com precise/restricted TranslationIndex
Hit http://archive.ubuntu.com precise/universe TranslationIndex
Get:10 http://archive.ubuntu.com precise-updates/main amd64 Packages [644 kB]
Get:11 http://security.ubuntu.com precise-security/restricted i386 Packages [4620 B]
Get:12 http://security.ubuntu.com precise-security/universe i386 Packages [79.8 kB]
Get:13 http://security.ubuntu.com precise-security/multiverse i386 Packages [2367 B]
Get:14 http://security.ubuntu.com precise-security/main TranslationIndex [74 B]
Get:15 http://security.ubuntu.com precise-security/multiverse TranslationIndex [71 B]
Get:16 http://security.ubuntu.com precise-security/restricted TranslationIndex [72 B]
Get:17 http://security.ubuntu.com precise-security/universe TranslationIndex [73 B]
Get:18 http://security.ubuntu.com precise-security/main Translation-en [143 kB]
Hit http://security.ubuntu.com precise-security/multiverse Translation-en
Get:19 http://security.ubuntu.com precise-security/restricted Translation-en [1253 B]
Get:20 http://security.ubuntu.com precise-security/universe Translation-en [49.1 kB]
Get:21 http://archive.ubuntu.com precise-updates/restricted amd64 Packages [10.1 kB]
Get:22 http://archive.ubuntu.com precise-updates/universe amd64 Packages [207 kB]
Get:23 http://archive.ubuntu.com precise-updates/multiverse amd64 Packages [13.6 kB]
Get:24 http://archive.ubuntu.com precise-updates/main i386 Packages [662 kB]
Get:25 http://archive.ubuntu.com precise-updates/restricted i386 Packages [10.0 kB]
Get:26 http://archive.ubuntu.com precise-updates/universe i386 Packages [211 kB]
Get:27 http://archive.ubuntu.com precise-updates/multiverse i386 Packages [13.8 kB]
Get:28 http://archive.ubuntu.com precise-updates/main TranslationIndex [3564 B]
Get:29 http://archive.ubuntu.com precise-updates/multiverse TranslationIndex [2605 B]
Get:30 http://archive.ubuntu.com precise-updates/restricted TranslationIndex [2461 B]
Get:31 http://archive.ubuntu.com precise-updates/universe TranslationIndex [2850 B]
Hit http://archive.ubuntu.com precise/main Translation-en
Hit http://archive.ubuntu.com precise/multiverse Translation-en
Hit http://archive.ubuntu.com precise/restricted Translation-en
Hit http://archive.ubuntu.com precise/universe Translation-en
Get:32 http://archive.ubuntu.com precise-updates/main Translation-en [296 kB]
Get:33 http://archive.ubuntu.com precise-updates/multiverse Translation-en [7834 B]
Get:34 http://archive.ubuntu.com precise-updates/restricted Translation-en [2432 B]
Get:35 http://archive.ubuntu.com precise-updates/universe Translation-en [122 kB]
Fetched 3280 kB in 8s (387 kB/s)       bzip2 0 B] [35 T
Reading package lists... Donege lists... 0%
Reading package lists... Donege lists... 0%
Building dependency tree... 66%ency tree... 0%
Reading state information... Doneormation... 0%
    amazon-ebs: E: Unable to locate package redis-server
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: Deleting temporary security group...
==> amazon-ebs: Deleting temporary keypair...
Build 'amazon-ebs' errored: Script exited with non-zero exit status: 100

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: Script exited with non-zero exit status: 100

==> Builds finished but no artifacts were created.
$
```

**Comment by mlococo** [comment]
Some thoughts on waiting for cloud-init more effectively than a hardcoded sleep:

- Sufficiently new cloud-init versions (0.7.5 and newer, I believe) have json status files in /var/lib/cloud/data/ and /run/cloud-init that are documented at http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/view/head:/doc/status.txt
- All cloud-init versions (or at least 0.6.3 and newer which shipped with Ubuntu 12.04) update /var/lib/cloud/instance/boot-finished. That file is persistent across boots, though, so it only works once. That might be ok for packer, but could lead to weirdness if it's left lying around in some upstream AMI. Documented in https://bugs.launchpad.net/cloud-init/+bug/1258113
- uvtool (which appears to be an Ubuntu-focused packer-like image creation tool) uses the following code to wait for cloud-init to finish: http://bazaar.launchpad.net/~uvtool-dev/uvtool/trunk/view/head:/remote-wait.sh

**Comment by arungkrishna** [comment]
> This was biting me, too, and I think I might know what's going on. On a typical Ubuntu EC2 instance, the top of `/etc/apt/sources.list` says:
> 
> ```
> ## Note, this file is written by cloud-init on first boot of an instance
> ## modifications made here will not survive a re-bundle.
> ```
> 
> Indeed `apt-get update` _should_ be using a mirror like `http://us-east-1.ec2.archive.ubuntu.com`. The normal sequence of events is:
> 
> 1. Instance boots up and the ssh service starts.
> 2. The cloud-init service starts and overwrites `/etc/apt/sources.list` with an EC2-specific mirror.
> 3. `sudo apt-get update` runs against the _new_ mirror.
> 4. `sudo apt-get install -y redis-server` runs against the _new_ mirror and everything works.
> 
> However, `apt-get update` is still using `http://archive.ubuntu.com` when the error happens, which makes me think the problematic sequence of events is:
> 
> 1. Instance boots up and the ssh service starts.
> 2. `sudo apt-get update` starts running against the _old_ mirror.
> 3. Meanwhile, the cloud-init service starts and overwrites `/etc/apt/sources.list` with an EC2-specific mirror.
> 4. `sudo apt-get install -y redis-server` runs against the _new_ mirror, but errors out because `apt-get update` has never been run against this particular mirror.
> 
> The workaround I'm using is:
> 
> ```
>     "provisioners": [{
>         "type": "shell",
>         "inline": [
>             "sleep 30",
>             "sudo apt-get update",
>             "sudo apt-get install -y redis-server"
>         ]
>     }]
> ```
> 
> I assume Packer considers the system "ready" as soon as sshd is listening? Ideally it would check for some kind of indication that the system has finished booting before it runs any commands, but I don't know a good way to do this on Ubuntu, let alone the general case of any Linux distribution. Which is why I'm using a `sleep` statement for now...

This was very helpful, thank you very much. We had the same issue while installing expect on Ubuntu using packer.
